File: This&that/_scripting_languages/Ruby/design_patterns/factory.rb
#
# Factory pattern
#     - The objective is to create an object
#


class ShapeFactory
    def get_shape(shape_type)
        if shape_type == nil
            return nil
        end 

        if shape_type == "circle"
            return Circle.new
        end

        if shape_type == "rectangle"
            return Rectangle.new
        end

        if shape_type == "square"
            return Square.new
        end

        return nil
    end
end

class Rectangle
    def draw
        puts 'Drawing a Rectangle!'
    end
end

class Square 
    def draw
        puts 'Drawing a Square!'
    end
end

class Circle 
    def draw
        puts 'Drawing a Circle!'
    end
end

#======= Usage:

shape_factory = ShapeFactory.new

shape1 = shape_factory.get_shape("circle")
shape1.draw # 'Drawing a Circle!'

shape2 = shape_factory.get_shape("rectangle")
shape2.draw # 'Drawing a Rectangle!'

shape3 = shape_factory.get_shape("square")
shape3.draw # 'Drawing a Square!' 

File: This&that/_scripting_languages/Javascript/_design_patterns/factory/factory.js
/**
 * factory pattern
 * The objective is to create an object
 *
 * the Shape interface: draw() 
 */
function ShapeFactory(){
    this.getShape = function(shapeType){
        if(shapeType == null){
            return null;
        }   

        if(shapeType == "circle"){
            return new Circle();
        } 

        if(shapeType == "rectangle"){
            return new Rectangle();
        }

        if(shapeType == "square"){
            return new Square();
        }

        return null; 
    }
}

function Rectangle(){
    this.draw = function(){
        console.log('Drawing a Rectangle!');
    }
}

function Square (){
    this.draw = function(){
        console.log('Drawing a Square!');
    }
}

function Circle (){
    this.draw = function(){
        console.log('Drawing a Circle!');
    }
}

//------- usage:

var shapeFactory = new ShapeFactory();

var shape1 = shapeFactory.getShape("circle");
shape1.draw();

var shape2 = shapeFactory.getShape("rectangle");
shape2.draw();

var shape3 = shapeFactory.getShape("square");
shape3.draw(); 


/* output:

Drawing a Circle!
Drawing a Rectangle!
Drawing a Square!
 
*/
